rfd/0211-azure-devops-joining.md

```diff
@@ -0,0 +1,486 @@
+IMK5zVF-jdILhng2RV_t0xm5pUUBN0gt7u-QKJ2px9iBdzYBJEtdR5Q-F
+ExwM7WDTZp4w2wDVslkaExAd_K3IoN4tqZk6FhKhsB3IRNMdsm9hGxwjt
+Jg3QkTETr8PZN2IUAKP6p7lXD1sXZOtV8exLh1VzTMqUHMqbWlVsG0rUX
+1WF7g"}
+```
+
+### ID Token Validation
+
+The ID Token is issued from an organization-specific issuer (e.g contains an
+`iss` claim that is specific to the organization).
+
+The OpenID configuration document available at the well-known URL for this `iss`
+is as follows:
+
+```json
+{
+  "issuer": "https://vstoken.dev.azure.com/0ca3ddd9-f0b0-4635-a98c-5866526961b6",
+  "jwks_uri": "https://vstoken.dev.azure.com/.well-known/jwks",
+  "subject_types_supported": [
+    "public",
+    "pairwise"
+  ],
+  "response_types_supported": [
+    "id_token"
+  ],
+  "claims_supported": [
+    "sub",
+    "aud",
+    "exp",
+    "iat",
+    "iss",
+    "nbf"
+  ],
+  "id_token_signing_alg_values_supported": [
+    "RS256"
+  ],
+  "scopes_supported": [
+    "openid"
+  ]
+}
+```
+
+Of note is the fact that the JWKS URI is not specific to each organization,
+this has a few ramifications:
+
+- The JWKS values could be cached across join token resources.
+- We must not assume that because an ID Token passes validation for a specific
+  well-known.
+
+Notably, the issued JWTs include the `kid` header and this `kid` field is also
+present within the JWKS. This allows the correct JWK to be selected from the 
+JWKS for validation.
+
+In order to be able to fetch the correct OIDC configuration document, we will
+require knowledge ahead of time of the organization UUID. **The user will
+configure this as part of the ProvisionToken specification**.
+
+During validation, we will expect the `aud` claim to be
+`api://AzureADTokenExchange`.
+
+For validation, we will leverage the `github.com/coreos/go-oidc/v3/oidc` package
+as is used for other OIDC join methods (e.g `bitbucket`).
+
+### Join RPC
+
+### UX
+
+#### Provision Token
+
+The ProvisionToken resource will be extended with new fields to support the 
+configuration of the Azure Devops join method.
+
+Like other join methods, a `spec.azure_devops` field will be introduced to hold
+configuration specific to the Azure DevOps join method. This field will include:
+
+- Configuration necessary to validate the ID Token:
+  - The UUID of the organization.
+- The allow rules that control which pipelines are permitted to authenticate.
+
+```protobuf
+// ProvisionTokenSpecV2AzureDevops contains the Azure Devops-specific
+// configuration. 
+message ProvisionTokenSpecV2AzureDevops {
+  message Rule {
+    // Sub also known as Subject is a string that roughly uniquely identifies
+    // the workload. Example:
+    // `p://my-organization/my-project/my-pipeline`
+    // Mapped from the `sub` claim.
+    string Sub = 1 [(gogoproto.jsontag) = "sub,omitempty"];
+    // The name of the AZDO project. Example:
+    // `my-project`.
+    // Mapped out of the `sub` claim.
+    string ProjectName = 2 [(gogoproto.jsontag) = "project_name,omitempty"];
```

Comment:
Technically the `ProjectName`/`PipelineName` fields do not exist directly as claims in the JWT, and can be extracted from the `sub` claim. I'm a little split on whether these should exist as fields when they do not exist as claims. I can see the positives around usability here but I worry if this poses a compatibility concern in the future if the shape of the claims within the token shift.

Alternatively, we could just provide globbing support for the `Sub` field.

WDYT?

Reply:
I'd personally be drawn towards keeping them as fields rather than globbing the `Sub` field because unless you have a very flat/small ADO setup, you'd probably always need to filter by at least the project, right?

Reply:
> I'd personally be drawn towards keeping them as fields rather than globbing the Sub field because unless you have a very flat/small ADO setup, you'd probably always need to filter by at least the project, right?

This aligns pretty well with my gut feeling - I think we can safely prioritise the UX here over accuracy.